data_model/src/account.rs

```diff
@@ -148,24 +147,28 @@ pub mod model {
     /// Condition which checks if the account has the right signatures.
     #[derive(
         Debug,
+        Default,
         Display,
         Clone,
         PartialEq,
         Eq,
         PartialOrd,
         Ord,
-        Constructor,
         Decode,
         Encode,
         Deserialize,
         Serialize,
         IntoSchema,
     )]
-    #[serde(transparent)]
-    #[repr(transparent)]
-    // SAFETY: `SignatureCheckCondition` has no trap representation in `EvalueatesTo<bool>`
-    #[ffi_type(unsafe {robust})]
-    pub struct SignatureCheckCondition(pub EvaluatesTo<bool>);
+    #[ffi_type(opaque)]
+    #[allow(clippy::enum_variant_names)]
+    pub enum SignatureCheckCondition {
+        #[default]
+        AnyAccountSignatures,
+        AllAccountSignatures,
+        #[display(fmt = "[{_0:?}]")]
+        AllCustomSignatures(Vec<PublicKey>),
+    }
```

Comment:
we should have a discussion which variants the `SignatureConditionCheck` enum should contain. @Erigara @Arjentix your thoughs?

Reply:
IMO we should have generalized form `N` out of `M` signatures (`AnyAccountSignatures` is `1` out of `M`, `AllAccountSignatures` is `M` out of `M`).

Also idk how useful `AllCustomSignatures` would be since set of pubic keys could be adjusted (through `Mint` and `Burn` and then `M` out of `M` is used).

Other than that i not sure which conditions we will need. 

Should be have `WASM` script for signature check condition (with access to transaction)?

Reply:
What I don't like in the "N out of M` model is that for account signatures the `M` can be changed, potentially making the previously set `SignatureCheckCondition` not what the user intended

Reply:
> What I don't like in the "N out of Mmodel is that for account signatures theMcan be changed, potentially making the previously setSignatureCheckCondition` not what the user intended

this makes a good point. What happens to the `SignatureCheckCondition` when a key is removed from `Account::signatories`. Should the change be reflected?

Reply:
Sounds quite reasonable

Reply:
Hm, currently the default is to allow the tx to pass if _any_ of the keys from account signatories are used, not _all_. Removing the _any_ option breaks some tests

Reply:
I would consider that a bug. Which tests are that?

Reply:
The `transaction_signed_by_new_signatory_of_account_should_pass` explicitly adds a new key to the signatory and tries to sign a transaction with it. 

There are other failed tests, but the failure may be transient, haven't looked at them deeper

Reply:
except that I would add custom signatures on top of `AnyAccountSignatures` and `AllAccountSignatures` not as an alternative

Reply:
Currently, I've implemented the following conditions:

1. require any of the `Account::signatories`
2. require all of the `Account::signatories`
3. require all of the `Account::signatories` + additional keys

Reply:
> Because IMHO we'll always want some of account signatories to be included

That's fair. I also think we can simplify by having only the version with the additional keys. With the empty additional keys list the logic would be equivalent. Also, the empty boxed slice [doesn't allocate](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=add8ce577f7c50575b0e1e3763249c79), so the perf should be the same

Reply:
> we don't need N out of M

I don't insist on N out of M. I agree that change in M is valid concern. I think main usecase for `N out of M` is some kind of majority voting (majority, supermajority) which we can add later.